Added documentation for new System.Net APIs released in 3.0

jozkee opened the pull request:
List of APIs documented:

- CookieCollection:

1. M:System.Net.CookieCollection.Clear
2. M:System.Net.CookieCollection.Contains(System.Net.Cookie)
3. M:System.Net.CookieCollection.Remove(System.Net.Cookie)
4. M:System.Net.CookieCollection.System#Collections#Generic#IEnumerable<System#Net#Cookie>#GetEnumerator

- FileWebRequest:

1. M:System.Net.FileWebRequest.GetRequestStreamAsync
2. M:System.Net.FileWebRequest.GetResponseAsync

- IPEndPoint:

1. M:System.Net.IPEndPoint.Parse(System.ReadOnlySpan{System.Char})
2. M:System.Net.IPEndPoint.Parse(System.String)
3. M:System.Net.IPEndPoint.TryParse(System.String,System.Net.IPEndPoint@)
4. M:System.Net.IPEndPoint.TryParse(System.ReadOnlySpan{System.Char},System.Net.IPEndPoint@)

Jozkee commented:
@davidsh @wfurt @krwq @scalablecory  since you are the System.Net owners, could you please take a look?

Jozkee commented:
@mairaw @rpetrusha

scalablecory commented on xml/System.Net/CookieCollection.xml:
```diff
@@ -291,10 +291,18 @@
-        <remarks>To be added.</remarks>
+        <param name="cookie">The <see cref="T:System.Net.Cookie" /> to locate in the <see cref="T:System.Net.CookieCollection" />.</param>
+        <summary>Determines whether the <paramref name="cookie" /> is in the <see cref="T:System.Net.CookieCollection" />.</summary>
+        <returns><see langword="true" /> if <paramref name="cookie" /> is found in the <see cref="T:System.Net.CookieCollection" />; otherwise, is <see langword="false" />.</returns>
```
> otherwise, is false

you've got stray "is"

scalablecory commented on xml/System.Net/CookieCollection.xml:
```diff
@@ -778,10 +786,20 @@
-        <remarks>To be added.</remarks>
+        <param name="cookie">The <see cref="T:System.Net.Cookie" /> to remove form the <see cref="T:System.Net.CookieCollection" />.</param>
+        <summary>Determines whether the <paramref name="cookie" /> is in the <see cref="T:System.Net.CookieCollection" />.</summary>
+        <returns><see langword="true" /> if item was successfully removed from the <see cref="T:System.Net.CookieCollection" />; otherwise, <see langword="false" />. This method also returns <see langword="false" /> if item is not found in the original collection.</returns>
```
> if item is not found in the original collection

Was this copied from `ICollection`? There is no "item" parameter here.

scalablecory commented on xml/System.Net/FileWebRequest.xml:
```diff
@@ -717,9 +717,29 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <summary>Returns a <see cref="T:System.IO.Stream" /> object for writing data to the file system resource as an asynchronous operation.</summary>
+        <returns>The Task object representing the asynchronous operation.</returns>
```
"Task" -> `<see cref="...Task"/>`.

carlossanlop replied:
@scalablecory @Jozkee there is a neat feature to help provide code suggestions that can be bulk merged:

```suggestion
        <returns>The <see cref="T:System.Threading.Tasks.Task" /> object representing the asynchronous operation.</returns>
```

carlossanlop replied:
Note: The comment is already outdated so you may ignore it.

scalablecory commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -568,10 +568,12 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="s">A <see cref="T:System.RedOnlySpan`1" /> that contains an IP endpoint in dotted-quad notation or network byte order for IPv4 and in colon-hexadecimal notation for IPv6.</param>
+        <summary>Converts an IP network endpoint (address and port) represented as a <see cref="T:System.RedOnlySpan`1" /> to an IPEndPoint instance.</summary>
```
RedOnlySpan -> ReadOnlySpan

scalablecory commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -568,10 +568,12 @@
+        <param name="s">A <see cref="T:System.RedOnlySpan`1" /> that contains an IP endpoint in dotted-quad notation or network byte order for IPv4 and in colon-hexadecimal notation for IPv6.</param>
+        <summary>Converts an IP network endpoint (address and port) represented as a <see cref="T:System.RedOnlySpan`1" /> to an IPEndPoint instance.</summary>
+        <returns>An <see cref="T:System.Net.IPEndPoint" /> instance.</returns>
         <remarks>To be added.</remarks>
```
Remarks still todo?

Jozkee replied:
Since overloads are generated in the same html, for this case and `TryParse` I think it would be redundant to have practically the same remarks twice.

carlossanlop replied:
Remarks are optional, @scalablecory

scalablecory commented on xml/System.Net/FileWebRequest.xml:
```diff
@@ -799,9 +819,22 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <summary>Returns a response to a file system request as an asynchronous operation.</summary>
+        <returns>The Task object representing the asynchronous operation.</returns>
```
"Task" -> `<see cref="...Task"/>`.

scalablecory commented on xml/System.Net/FileWebRequest.xml:
```diff
@@ -717,9 +717,29 @@
+          <format type="text/markdown"><![CDATA[  
+          ## Remarks  
+          > [!NOTE]
+          >  In the case of asynchronous requests, it is the responsibility of the client application to implement its own time-out mechanism.
```
Should this be a link to something describing how to implement time-outs?

Jozkee replied:
there is an example for [`HttpWebRequest.BeginGetResponse`](https://docs.microsoft.com/en-us/dotnet/api/system.net.httpwebrequest.begingetresponse?view=netcore-3.0) I could adapt it to `FileWebRequest.GetResponseAsync` in case is really needed.

Jozkee replied:
Created issue #2672 to address this later.

mairaw commented:
You can see when your changes on the staging site: https://review.docs.microsoft.com/en-us/dotnet/api/system.net.ipendpoint?view=netcore-3.0&branch=pr-en-us-2663

mairaw commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -602,10 +604,30 @@
+|2 -- "20.65535:23"|20.0.255.255|23|  
+|3 -- "128.1.2:443"|128.1.0.2|443|  
+
+Note that this method accepts as valid an string value that can be parsed as an Int64, and then treats that Int64 as the long value of an IP address in network byte order, similar to the way that the IPAddress constructor does. This means that this method returns true if the Int64 is parsed successfully, even if it represents an address that's not a valid IP address. For example, if s is "1", thi ...
```
This seems like an important note.

```suggestion
> [!IMPORTANT]
> Note that this method accepts as valid an string value that can be parsed as an <xref:System.Int64>, and then treats that <xref:System.Int64> as the long value of an IP address in network byte order, similar to the way that the <xref:System.Net.IPAddress> constructor does. This means that this method returns `true` when the <xref:System.Int64> is parsed successfully, even if it represents an address that's not a valid IP address. For example, if `s` is "1", this method returns `true` even though "1" (or 0.0.0.1) is not a valid IP address. Fixing this bug would break existing apps, so the current behavior will not be changed. Your code can avoid this behavior by ensuring that it only uses this method to parse IP addresses in dotted-decimal format.
```

mairaw commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -793,10 +815,10 @@
+        <param name="s">The IP endpoint to validate.</param>
+        <param name="result">The <see cref="T:System.Net.IPEndPoint" /> version of <paramref name="s" />.</param>
+        <summary>Determines whether <paramref name="s" /> is a valid IP endpoint.</summary>
+        <returns><see langword="true" /> if <paramref name="s" /> was able to be parsed as an IP endpoint; otherwise, <see langword="false" />.</returns>
```
```suggestion
        <returns><see langword="true" /> if <paramref name="s" /> can be parsed as an IP endpoint; otherwise, <see langword="false" />.</returns>
```

mairaw commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -829,11 +851,29 @@
+        <param name="s">The IP endpoint to validate.</param>
+        <param name="result">The <see cref="T:System.Net.IPEndPoint" /> version of <paramref name="s" />.</param>
+        <summary>Determines whether <paramref name="s" /> is a valid IP endpoint.</summary>
+        <returns><see langword="true" /> if <paramref name="s" /> was able to be parsed as an IP endpoint; otherwise, <see langword="false" />.</returns>
```
```suggestion
        <returns><see langword="true" /> if <paramref name="s" /> can be parsed as an IP endpoint; otherwise, <see langword="false" />.</returns>
```

mairaw commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -829,11 +851,29 @@
+          <format type="text/markdown"><![CDATA[  
+  
+## Remarks  
+The number of parts (each part is separated by a period) in the string s determines how the endpoint's IP address is constructed. A one part address is stored directly in the network address. A two part address, convenient for specifying a class A address, puts the leading part in the first byte and the trailing part in the right-most three bytes of the network address. A three part address, conv ...
```
same comments as earlier for this set of remarks.

davidsh commented:
All of these types and methods aren't really new.  What's new is that there are some overloads with ReadOnlySpan<T> instead of other buffer types. 

So, I would advise looking at that existing documentation to re-use most of the same descriptions of parameters, remarks, return values, etc. We only need small changes when new parameter types such as the ReadOnlySpan<T> are added.

carlossanlop commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -764,6 +806,44 @@
       </Docs>
     </Member>
+    <MemberGroup MemberName="TryParse">
+      <AssemblyInfo>
```
Why did you add `AssemblyInfo` items manually? These items are supposed to be auto-generated by the Docs repo.

carlossanlop replied:
Same question about the other `AssemblyInfo` items in other sections of the PR.

Jozkee replied:
Because this is a `MemberGroup` which is not auto-generated. I also did it based on suggested example in https://github.com/dotnet/dotnet-api-docs/pull/2663#discussion_r298781159, which adds the `MemberGroup` node with an `AssemblyInfo` inside. 

But I think you are right, I removed the `AssemblyInfo` nodes and I was able to compile locally without problems, however I would like to have validation from @mairaw or @rpetrusha about this.

mairaw replied:
Pinging @joelmartinez on this one. I also added this manually to other member groups just because others I saw had it.

Joel, is it okay to remove `AssemblyInfo` from the MemberGroup elements?

joelmartinez replied:
@mairaw, yes, that is fine ... the membergroup, from a metadata perspective, is only informative; so it doesn't really conform to any of the other requirements reserved for members :)

carlossanlop commented on xml/System.Net/IPEndPoint.xml:
```diff
@@ -602,10 +642,12 @@
-        <returns>To be added.</returns>
+        <param name="s">A <see cref="T:System.String" /> that contains an IP endpoint in dotted-quad notation or network byte order for IPv4 and in colon-hexadecimal notation for IPv6.</param>
+        <summary>Converts an IP network endpoint (address and port) represented as a <see cref="T:System.String" /> to an <see cref="T:System.Net.IPEndPoint" /> instance.</summary>
+        <returns>The object representation of an IP network endpoint</returns>
```
```suggestion
        <returns>The object representation of an IP network endpoint.</returns>
```

Jozkee commented:
This is ready for re-review:
* Added missing dot.
* Removed `AssemblyInfo` from `MemberGroup`s.
* As per https://github.com/dotnet/dotnet-api-docs/pull/2767#discussion_r304644461, first exception conditions should be written in the same line than `exception` tag.

@mairaw @rpetrusha

Jozkee commented:
@mairaw, @rpetrusha, Its been a while since I've addressed the comments on this one, can we merge it?

mairaw commented:
Taking a look now @Jozkee!

mairaw commented:
@Jozkee I've made some changes to the three files. Can you make sure I didn't introduce anything wrong? Also let me know if you dislike any of the changes I made. Thanks!

Jozkee commented on xml/System.Net/CookieCollection.xml:
```diff
@@ -860,9 +879,20 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <summary>Gets an enumerator that can iterate through the <see cref="T:System.Net.CookieCollection" />.</summary>
+        <returns>An <see cref="T:System.Collections.Generic.IEnumerator`1" /> for this collection.</returns>
```
Same as GetEnumerator in line 527.
```suggestion
        <returns>A generic enumerator for this collection.</returns>
```

mairaw replied:
I think we can keep this one as-is

mairaw commented on xml/System.Net/CookieCollection.xml:
```diff
@@ -515,7 +524,7 @@
       <Docs>
         <summary>Gets an enumerator that can iterate through a <see cref="T:System.Net.CookieCollection" />.</summary>
-        <returns>An instance of an implementation of an <see cref="T:System.Collections.IEnumerator" /> interface that can iterate through a <see cref="T:System.Net.CookieCollection" />.</returns>
+        <returns>An <see cref="T:System.Collections.Generic.IEnumerator" /> for this collection.</returns>
```
I've added to match what we had for similar APIs. I just simplified the sentence a bit, did not add the cref.

Jozkee commented:
@mairaw Looks good to me, can we merge?

carlossanlop commented:
@mairaw thanks for your last commit to fix the broken cref. The latest build passed without warnings. Is this good to merge?

carlossanlop commented:
@mairaw successful build with no warnings. I think it's ready to merge.
